### PR TITLE
Issue 1361: GetScalingEvents REST API

### LINKED ISSUE #1361: Implement REST API call to return the list of scale events
Subtask of #1347 .

### PR DESCRIPTION
**Pull request by Abhijeet-Jadhav**
**Change log description**
Addition of new REST API - getScalingEvents. Returns a list of scaling events in desired period of time.

**Purpose of the change**
Fixes #1361

**What the code does**
Adds a new REST API implementation.

**How to verify it**
Unit Test/ Call the api using a rest client by starting standalone pravega cluster.

Note: Please do not review auto-generated files /rest/generated/*.

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 1]
```diff
@@ -407,4 +412,51 @@ public void updateStreamState(final String scopeName, final String streamName,
...
+     * @param from              DateTime from which scaling events should be displayed.
+     * @param to                DateTime until which scaling events should be displayed.
+     * @param securityContext   The security for API access.
+     * @param asyncResponse     AsyncResponse provides means for asynchronous server side response processing.
+     */
+    @Override
+    public void getScalingEvents(final String scopeName, final String streamName, final Long from, final Long to,
+                                 final SecurityContext securityContext, final AsyncResponse asyncResponse) {
+        long traceId = LoggerHelpers.traceEnter(log, "getScalingEvents");
+
```
Comment (skrishnappa) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Add input validation for from and to params (like >=0, from < to, etc) and return 400 on error.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 2]
```diff
@@ -407,4 +412,51 @@ public void updateStreamState(final String scopeName, final String streamName,
...
+        controllerService.getScaleRecords(scopeName, streamName).thenApply(listScaleMetadata -> {
```
Comment (skrishnappa) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: We are getting all records and filtering everytime, can we instead pass on the time range params into the controllerservice function and filter at the store layer? only if its possible to filter internally to save I/O operations.
Also if this list can potentially be large, then we may need some limit or paging support, file an issue for this if this is a valid case.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 3]
```diff
@@ -407,4 +412,51 @@ public void updateStreamState(final String scopeName, final String streamName,
...
+            Iterator<ScaleMetadata> metadataIterator = listScaleMetadata.iterator();
+            List<ScaleMetadata> finalScaleMetadataList = new ArrayList<ScaleMetadata>();
+            ScaleMetadata referenceEvent = null;
+            while (metadataIterator.hasNext()) {
+                ScaleMetadata scaleMetadata = metadataIterator.next();
+                if (scaleMetadata.getTimestamp() >= from && scaleMetadata.getTimestamp() <= to) {
+                    finalScaleMetadataList.add(scaleMetadata);
+                } else if (scaleMetadata.getTimestamp() < from && scaleMetadata.getTimestamp() < to) {
```
Comment (skrishnappa) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Why this check and this event? Also please add comment in the code

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 4]
```diff
@@ -407,4 +412,51 @@ public void updateStreamState(final String scopeName, final String streamName,
...
+            return Response.status(Status.OK).entity(finalScaleMetadataList).build();
+        }).exceptionally(exception -> {
+            if (exception.getCause() instanceof DataNotFoundException
+                    || exception instanceof DataNotFoundException
+                    || exception.getCause() instanceof StoreException.NodeNotFoundException
+                    || exception instanceof StoreException.NodeNotFoundException) {
+                log.warn("Stream/Scope name: {}/{} not found", scopeName, streamName);
+                return Response.status(Status.NOT_FOUND).build();
+            } else {
+                log.warn("getScalingEvents for {} failed with exception: {}", scopeName, exception);
```
Comment (skrishnappa) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Include streamName and also last {} not required.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 5]
```diff
@@ -407,4 +412,51 @@ public void updateStreamState(final String scopeName, final String streamName,
...
+    @Override
+    public void getScalingEvents(final String scopeName, final String streamName, final Long from, final Long to,
+                                 final SecurityContext securityContext, final AsyncResponse asyncResponse) {
+        long traceId = LoggerHelpers.traceEnter(log, "getScalingEvents");
+
+        controllerService.getScaleRecords(scopeName, streamName).thenApply(listScaleMetadata -> {
+            Iterator<ScaleMetadata> metadataIterator = listScaleMetadata.iterator();
+            List<ScaleMetadata> finalScaleMetadataList = new ArrayList<ScaleMetadata>();
+            ScaleMetadata referenceEvent = null;
+            while (metadataIterator.hasNext()) {
```
Comment (skrishnappa) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: could be simplified by using forEach?

File: shared/controller-api/src/main/swagger/Controller.yaml [hunk 6]
```diff
@@ -315,3 +360,29 @@ definitions:
...
+      segmentList:
+        type: array
+        items:
+          $ref: "#/definitions/Segment"
+  Segment:
+    type: object
+    properties:
+      number:
+        type: integer
+        format: int34
```
Comment (skrishnappa) [inline shared/controller-api/src/main/swagger/Controller.yaml]: I guess should be int32

File: shared/controller-api/src/main/swagger/Controller.yaml [hunk 7]
```diff
@@ -315,3 +360,29 @@ definitions:
...
+      start:
```
Comment (skrishnappa) [inline shared/controller-api/src/main/swagger/Controller.yaml]: startTime?

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 8]
```diff
@@ -407,4 +412,51 @@ public void updateStreamState(final String scopeName, final String streamName,
...
+                    || exception.getCause() instanceof StoreException.NodeNotFoundException
+                    || exception instanceof StoreException.NodeNotFoundException) {
+                log.warn("Stream/Scope name: {}/{} not found", scopeName, streamName);
+                return Response.status(Status.NOT_FOUND).build();
+            } else {
+                log.warn("getScalingEvents for {} failed with exception: {}", scopeName, exception);
+                return Response.status(Status.INTERNAL_SERVER_ERROR).build();
+            }
+        }).thenApply(asyncResponse::resume)
+                .thenAccept(x -> LoggerHelpers.traceLeave(log, "getScalingEvents", traceId));
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: The indentation here seems incorrect.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 9]
```diff
@@ -407,4 +412,51 @@ public void updateStreamState(final String scopeName, final String streamName,
...
+                                 final SecurityContext securityContext, final AsyncResponse asyncResponse) {
+        long traceId = LoggerHelpers.traceEnter(log, "getScalingEvents");
+
+        controllerService.getScaleRecords(scopeName, streamName).thenApply(listScaleMetadata -> {
+            Iterator<ScaleMetadata> metadataIterator = listScaleMetadata.iterator();
+            List<ScaleMetadata> finalScaleMetadataList = new ArrayList<ScaleMetadata>();
+            ScaleMetadata referenceEvent = null;
+            while (metadataIterator.hasNext()) {
+                ScaleMetadata scaleMetadata = metadataIterator.next();
+                if (scaleMetadata.getTimestamp() >= from && scaleMetadata.getTimestamp() <= to) {
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Does it mean that `getScaleRecords` returns all events from the beginning of times? If a stream has been going on for a while, the list might be long and might want to garbage collect info on the events early on. This is out of scope for this pull request, though. I'm pointing it out here because I saw the from/to variables.

File: controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java [hunk 10]
```diff
@@ -642,6 +644,52 @@ public void testUpdateStreamState() throws Exception {
...
+        Response response = client.target(resourceURI).queryParam("from", fromDateTime).
+                queryParam("to", toDateTime).request().buildGet().invoke();
+        assertEquals("Get Scaling Events response code", 200, response.getStatus());
+        final List<ScaleMetadata> scaleMetadataListResponse = response.readEntity(List.class);
+        assertEquals("List Size", 3, scaleMetadataListResponse.size());
+
+        // Test for getScalingEvents for invalid scope.
+        final CompletableFuture<List<ScaleMetadata>> completableFuture1 = new CompletableFuture<>();
+        completableFuture1.completeExceptionally(new DataNotFoundException(""));
+        when(mockControllerService.getScaleRecords("scope1", "stream1")).thenReturn(completableFuture1);
```
Comment (fpj) [inline controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java]: I think we are assuming that data not found means invalid scope, yes? because otherwise, `scope1` seems like a valid one. What if the stream doesn't exist? Is there any other scenario in which I can get a data not found, which is not via invalid scope?

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 1 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Done.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 5 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Cannot use non-final variable like referenceEvent in forEach hence iterator.
(Variable used in lambda expression should be final or effectively final)

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 3 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: This check is required to record refernceEvent for comparison of events between 'from' and 'to'. Added comments.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 4 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Done.

File: shared/controller-api/src/main/swagger/Controller.yaml (hunk 6 above)
Comment (Abhijeet-Jadhav) [inline shared/controller-api/src/main/swagger/Controller.yaml]: yes it should be. Done.

File: shared/controller-api/src/main/swagger/Controller.yaml (hunk 7 above)
Comment (Abhijeet-Jadhav) [inline shared/controller-api/src/main/swagger/Controller.yaml]: changed to startTime.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 8 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Done.

File: controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java (hunk 10 above)
Comment (Abhijeet-Jadhav) [inline controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java]: DataNotFound in this case is for stream or scope not found. Non-existent stream is covered in this case. This is consistent with other rest apis.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 9 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Yes, getScaleRecords returns all events from the beginning of times. Discussing offline for optimizations to this call, with @skrishnappa and @shiveshr.

File: shared/controller-api/src/main/swagger/Controller.yaml [hunk 11]
```diff
@@ -284,8 +284,53 @@ paths:
...
+        - application/json
+      responses:
+        200:
+          description: List of scaling events
+          schema:
+            $ref: "#/definitions/ScalingEventList"
+        404:
+          description: Scope/Stream not found
+        500:
+          description: Server error
```
Comment (skrishnappa) [inline shared/controller-api/src/main/swagger/Controller.yaml]: Expand message to be more descriptive since we will be generating public document from this (see other description after merging with master).

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 3 above)
Comment (skrishnappa) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: From the requirement here - https://github.com/pravega/pravega/issues/1347, seems like we need the initial state of the stream -
"Should return a “scaling event” that describes the initial state of the stream when created. Will be the only entry returned for static segment streams"
But here we might be returning any arbitrary event. Also please review if this event is guaranateed to be present, cc @shiveshr.

Remove the redundant "< to" check

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+public class Segment   {
+  private Integer number = null;
+
+  private Long startTime = null;
+
+  private Integer keyStart = null;
+
+  private Integer keyEnd = null;
+
+  public Segment number(Integer number) {
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java]: you can use builder pattern here

File: controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java (hunk 10 above)
Comment (shiveshr) [inline controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java]: DataNotFound can occur if any node in the hierarchy does not exists. 
the hierarchy is scope/stream/stream-metadata
So for this api call if segmentTable doesnt exist or history table doesnt exist we can get DataNotFoundException.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScaleMetadata.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+import java.util.Objects;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import io.pravega.controller.server.rest.generated.model.Segment;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/ScaleMetadata.java]: swagger auto-generated file.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+
+
+
+/**
+ * Segment
+ */
+
+public class Segment   {
+  private Integer number = null;
```
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java]: swagger framework's auto-generated file.

File: controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java (hunk 12 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/generated/model/Segment.java]: swagger framework's auto-generated file.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 2 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Issue filed for paging support https://github.com/pravega/pravega/issues/1463

File: shared/controller-api/src/main/swagger/Controller.yaml (hunk 11 above)
Comment (Abhijeet-Jadhav) [inline shared/controller-api/src/main/swagger/Controller.yaml]: Done.

File: controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java (hunk 10 above)
Comment (Abhijeet-Jadhav) [inline controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java]: In that case, what we return `404 Not Found` is appropriate.

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 3 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: removed the redundant "< to" checked.

Comment (skrishnappa) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: cc @glounder

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java [hunk 15]
```diff
@@ -405,6 +410,67 @@ public void updateStreamState(final String scopeName, final String streamName,
...
+     */
+    @Override
+    public void getScalingEvents(final String scopeName, final String streamName, final Long from, final Long to,
+                                 final SecurityContext securityContext, final AsyncResponse asyncResponse) {
+        long traceId = LoggerHelpers.traceEnter(log, "getScalingEvents");
+
+        if (from < 0 || to < 0 || from > to) {
+            log.warn("Received invalid request from client for scopeName/streamName: {}/{} ", scopeName, streamName);
+            asyncResponse.resume(Response.status(Status.BAD_REQUEST).build());
+            return;
```
Comment (skrishnappa) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Add traceLeave before return

File: controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java [hunk 16]
```diff
@@ -642,6 +644,60 @@ public void testUpdateStreamState() throws Exception {
...
+
+        when(mockControllerService.getScaleRecords(scope1, stream1)).
+                thenReturn(CompletableFuture.completedFuture(scaleMetadataList));
+        Response response = client.target(resourceURI).queryParam("from", fromDateTime).
+                queryParam("to", toDateTime).request().buildGet().invoke();
+        assertEquals("Get Scaling Events response code", 200, response.getStatus());
+        final List<ScaleMetadata> scaleMetadataListResponse = response.readEntity(List.class);
+        assertEquals("List Size", 2, scaleMetadataListResponse.size());
+
+        // Test for getScalingEvents for invalid scope.
```
Comment (skrishnappa) [inline controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java]: Can we add a test case for static stream segments, i.e. streams without any scale events. Do we expect an event with the stream creation info for this?

File: controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java (hunk 15 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java]: Done.

File: controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java (hunk 16 above)
Comment (Abhijeet-Jadhav) [inline controller/src/test/java/io/pravega/controller/rest/v1/StreamMetaDataTests.java]: Mocked the segments info here to simulate it and for the actual scenario the first event will be the initial segment info.
